distributed/shuffle/_disk.py

```diff
@@ -89,6 +93,138 @@ def read(self) -> Generator[None, None, None]:
             self.release_read()
 
 
+class NewDiskBuffer:
+    def __init__(self, directory, memory_limiter: ResourceLimiter):
+        self.memory_limiter = memory_limiter
+        self._directory_lock = ReadWriteLock()
+        self._closed = False
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self.bytes_total = 0
+        self.bytes_memory = 0
+        self.bytes_written = 0
+        self.bytes_read = 0
+        self._exception = None
+        self._accepts_input = True
+        self._inputs_done = False
+        self._flush_lock = asyncio.Lock()
+        from dask.utils import parse_bytes
+
+        self._write_lock = asyncio.Lock()
+        self.min_size = parse_bytes("1MB")
+        self.max_size = parse_bytes("500MB")
+        self.directory = pathlib.Path(directory)
+        self.directory.mkdir(exist_ok=True)
+
+    def heartbeat(self) -> dict[str, Any]:
+        return {
+            "memory": self.bytes_memory,
+            "total": self.bytes_total,
+            "buckets": len(self.shards),
+            "written": self.bytes_written,
+            "read": self.bytes_read,
+            "diagnostics": {},
+            "memory_limit": self.memory_limiter.limit,
+        }
+
+    async def write(self, data):
+        if self._exception:
+            raise self._exception
+        if not self._accepts_input or self._inputs_done:
+            raise RuntimeError(f"Trying to put data in closed {self}.")
+
+        if not data:
+            return
+
+        sizes = {bucket: shard.nbytes for bucket, shard in data.items()}
+        total_batch_size = sum(sizes.values())
+        self.bytes_memory += total_batch_size
+        self.bytes_total += total_batch_size
+
+        self.memory_limiter.increase(total_batch_size)
+        async with self._write_lock:
+            for worker, shard in data.items():
+                self.shards[worker].append(shard)
+                self.sizes[worker] += sizes[worker]
+            del data
+            assert total_batch_size
+            while self.memory_limiter.full or sum(self.sizes.values()) > self.max_size:
+                await self.flush_one()
+
+    def write_data_to_disk(self, data: Iterable[pa.Table], bucket: int) -> None:
+        import pyarrow as pa
+
+        tab = pa.concat_tables(data)
+        with self._directory_lock.read():
+            if self._closed:
+                raise RuntimeError("Already closed")
+            with pa.ipc.new_file(self.directory / str(bucket), tab.schema) as f:
+                f.write_table(tab)
```

Comment:
From what I can tell, this does not work as intended. It overwrites the contents of an existing file instead of appending. Tests seem to work because we never reach the limit of `1 GiB` for disk. If you reduce this to `1 MiB`, the test suite starts failing.

Reply:
I already suspected this and was surprised to see tests work. I think I saw some data loss in tests at some point, though. Anyhow, there is a reason why this is a draft :)